Simplify dev setup

hwbllmnn opened the pull request:
This simplifies the dev setup and documents the necessary steps to run the dev setup in the `README.md`.

This is done by running a container for all components, including the api, and configuring the `nginx` to reverse proxy the api and Geoserver. For the api container, a `Dockerfile.dev` is added that mounts the repo into the container and runs the flask dev server.

This has the following advantages:

* in order to resume work later on, one just has to run `docker compose up` to start the whole system
* in preparation for a test or prod environment, the only change necessary would be to switch the api container to a prod build with gunicorn, and possibly change the `nginx` `default.conf` to reverse proxy to a different port

The latter point is especially important to stay as close to the prod environment as possible when developing.

Since the repo is mounted into the api container, you can just make changes locally and they will be used by the dev server immediately, making it no different than running flask locally. Also, there's no need any more to have all those python packages installed locally.

What do you think?

StefanSchuhart commented:
Just a few notes to make sure we all have a basic understanding of our development styles and needs, before reviewing your proposal: I, personally, prefer to use dev containers that contain all tools I need for my development purposes. These I can move from one development machine to another, without the need to build the whole setup from scratch. From there I use local, project-related virtual python environments (.venv/). 
Inside this dev container I run the code I build. Additionally, all dependent software like databases have their own containers specified in docker-compose-dev.yaml. This has the advantage, among other things, that I`m developing inside a docker managed network and can use the containers names as domain names, without the need for nginx and a special configuration. Also, in production we do not use nginx anymore, instead we use traefik as it feels more naturally integrated into container based production environments. 

Considering the changes in launch.json: Please add your preferred configuration instead of overriding the existing one.

StefanSchuhart commented:
I would suggest two changes: Would you mind copy your dev setup from docker-compose.yaml to docker-compose-dev.yaml? Docker-compose.yaml contains the current production setup including the UMP version currently in production. Could you switch your pull request from main branch to dev using --no-ff flag? This enables us to select what goes into the main branch.

hwbllmnn commented:
Alright, we've rebased against `dev` and reset the configurations as they were before.

We now added an alternate `docker-compose-local.yaml` for those people who'd like to start everything with a single compose command, so that it does not interfere with the previous dev setup. We've also updated the `README.md` to reflect the new approach.
